### PR TITLE
BaseTools/GenFw X64: Drop GOT handling

### PR DESCRIPTION
**Pull request by ardbiesheuvel**
Modern x86_64 ELF toolchains rarely emit GOT entries when generating fully linked binaries such as the ones used by GenFw for ELF-to-PECOFF conversion. This is not only due to LTO, where the linker can see all symbols in the program and decide that GOT indirections are unnecessary; even when not using LTO, the compiler and linker can rely on relaxable GOTPCRELX relocations to obtain the same result. These relocations are described in the ELF psABI for x86_64, and can be relaxed to direct symbol references by the linker if they can be satisfied by symbol defined in the same executable (as opposed to ones defined in shared libraries).

GenFw currently implements support for GOT based relocations on x86_64, but this is tricky code* that has to work around the fact that no static relocations are emitted for the GOT by the linker, only dynamic ones. This is due to the fact that the --emit-relocs linker switch it relies on emits the relocations generated by the compiler, whereas the GOT is constructed by the linker itself.

It also assumes that linker relaxations are reflected in the output of --emit-relocs. For instance, when relaxing

  movq  foo@GOTPCREL(%rip), %rax

to

  leaq  foo(%rip), %rax

(which loads the address of 'foo' into RAX), the resulting GOTPCRELX relocation must either be suppressed or converted into a R_X86_64_PC32 relocation, or the reference will be misidentified.  When this happens, GenFw will emit a PE/COFF relocation that refers to 'foo' directly rather than the GOT slot carrying its absolute address, resulting in corruption of the code at the symbol's address when the PE/COFF loader applies the relocation to what it assumes is an absolute address that needs to be relocated.

ld.bfd and LLD behave differently in this regard, and given that --emit-relocs is a crutch to begin with, it is not obvious what should be emitted for the case above.

If GOT support is really needed in GenFw, it would be *much* better to implement this based on dynamic relocations which are emitted when linking Position Independent Executables (PIE). PIE binaries are built to tolerate placement anywhere in the 64-bit address space, and its relocation section only describes absolute quantities that require handling by GenFw.

However, it appears we don't actually need to handle GOTs at all.  So remove support for X64 GOT handling from GenFw:
- remove the .got input sections from the linker scripts
- remove GOTPCREL handling, so that non-relaxable GOTPCREL relocations are rejected by GenFw;
- ignore GOTPCRELX relocations, which will be relaxed by the linker, and not cause an entry to be emitted into the .got section.

*case in point: it needs a 'safety net'

### CODE REVIEW

**Comment by ardbiesheuvel** [comment]
@kraxel @mikebeaton

**Comment by mikebeaton** [comment]
This gives the same results for me as your [test fix](https://github.com/tianocore/edk2/pull/6346#issuecomment-2424986597): works perfectly, with no increase in NOOPT CLANGDWARF X64 build size; does not cause any issues building and running OvmfPkg for other builds (DEBUG, RELEASE) or toolchains (CLANGPDB, GCC, XCODE5) (VS2019 not tested, but assume CI will catch this).

**Comment by mikebeaton** [comment]
If we need to keep .got input sections in GCC (to make other architectures such as LOONGARCH64 build), does it make sense to lose them in CLANG? (E.g. might CLANG be used to build architectures other than X64 which still need GOT sections?) Even if we do need to lose them in CLANG but keep them in GCC, it might good to have a comment explaining how the fact of having them in one but not the otherinteracts with the shared Elf64Convert code?

**Comment by mikebeaton** [comment]
> If we need to keep .got input sections in GCC (to make other architectures such as LOONGARCH64 build), does it make sense to lose them in CLANG? (E.g. might CLANG be used to build architectures other than X64 which still need GOT sections?) Even if we do need to lose them in CLANG but keep them in GCC, it might be helpful to have an additional comment in the commit message explaining how having them in one but not the other interacts with the shared Elf64Convert code?

Addressed, tyvm. I can retest, though may be not be for some hours from now.

**Comment by mikebeaton** [comment]
Based on this commit https://github.com/tianocore/edk2/commit/01225075dbf my understanding is that GCC would become the rolling current toolchain, and GCC5 would possibly stay as is and become legacy at some point? cc @bexcran 

If so, again in my understanding, and apologies if I'm missing something, the updates here would ideally be applied to both GCC and GCC5?

File: BaseTools/Conf/tools_def.template [hunk 1]
```diff
@@ -988,9 +988,9 @@ DEFINE GCC5_X64_CC_FLAGS             = DEF(GCC49_X64_CC_FLAGS)
 DEFINE GCC5_IA32_X64_DLINK_COMMON    = DEF(GCC49_IA32_X64_DLINK_COMMON)
 DEFINE GCC5_IA32_X64_ASLDLINK_FLAGS  = DEF(GCC49_IA32_X64_ASLDLINK_FLAGS)
 DEFINE GCC5_IA32_X64_DLINK_FLAGS     = DEF(GCC49_IA32_X64_DLINK_FLAGS)
-DEFINE GCC5_IA32_DLINK2_FLAGS        = DEF(GCC49_IA32_DLINK2_FLAGS) -Wno-error
+DEFINE GCC5_IA32_DLINK2_FLAGS        = DEF(GCC49_IA32_DLINK2_FLAGS) -Wno-error -Wl,-u,gotassert
 DEFINE GCC5_X64_DLINK_FLAGS          = DEF(GCC49_X64_DLINK_FLAGS)
-DEFINE GCC5_X64_DLINK2_FLAGS         = DEF(GCC49_X64_DLINK2_FLAGS) -Wno-error
+DEFINE GCC5_X64_DLINK2_FLAGS         = DEF(GCC49_X64_DLINK2_FLAGS) -Wno-error -Wl,-u,gotassert
```
Comment (mikebeaton) [inline BaseTools/Conf/tools_def.template:993]: Based on this commit https://github.com/tianocore/edk2/commit/01225075dbf my understanding is that GCC would become the rolling current toolchain, and GCC5 would possibly stay as is and become legacy at some point? cc @bexcran 

If so, again in my understanding and apologies if I'm missing something, the updates here would ideally be applied to both GCC and GCC5?

Comment (ardbiesheuvel) [reply BaseTools/Conf/tools_def.template:993]: Yes. Counter-intuitively, `GCC5_X64_DLINK2_FLAGS` is actually consumed to construct `*_GCC_X64_DLINK2_FLAGS` so this should work as expected.

Comment (mikebeaton) [reply BaseTools/Conf/tools_def.template:993]: Ahh, right! I don't think I have an option to mark as resolved, but resolved, tyvm.

**Comment by ardbiesheuvel** [comment]
Unfortunately, I am going to have to drop these patches, as it turns out there are cases where the x86_64 GCC toolchain still emits GOT entries.

E.g., when building CryptoPei, I see some references such as

```
   4e215:       48 3b 05 74 f2 13 00    cmp    0x13f274(%rip),%rax        # 18d490 <.got>
                        4e218: R_X86_64_REX_GOTPCRELX   aesni_ctr32_encrypt_blocks-0x4
```

where a pointer is compared to the address of the function `aesni_ctr32_encrypt_blocks()`, and this function is implemented in assembler and so its definition is not visible to the LTO linker.

Relocations of this type are not relaxable under the PIE code model, and so they must be processed by `GenFw` so that the correct address is emitted into the GOT slot.

The consequence of this is that `GenFw` will need to reason about whether relocations of type `R_X86_64_REX_GOTPCRELX` have been relaxed or not. Only `movq` and `call` GOTPCREL relocations are relaxable under the PIE model, and only the former occur in practice in our configuration.

IOW, GenFw should disregard `R_X86_64_REX_GOTPCRELX` relocations that refer to `leaq` instructions, but handle all others as before.

**Pull request closed without merging**